adding a blog post for "percy magic" (#284)
simplabs/simplabs.github.io: _posts/2020-03-23-porting-a-site-to-ember-with-percy.md

---
title: 'Porting a site to Ember using Percy'
author: 'Chris Manson'
github: mansona
twitter: real_ate
topic: ember
bio: 'Senior Ember Devleoper'

> description: 'something something'

Reviewer: This is still missing

Author: I've added it 👍 let me know if you would like any changes to it

og:
  image: /assets/images/posts/2020-03-23-porting-a-site-to-ember-with-percy/og-image.png
---

This article details a technique where the visual testing software
[Percy](https://percy.io/) was used to port the old [Ember
Guides](https://guides.emberjs.com/release/) to an entirely different

> technology, without anybody noticing

Reviewer: ```suggestion
technology, without anybody noticing.
```

<!--break-->

> This article details one of the first steps in the long road to _Emberify_ and

Reviewer: Can this paragraph start with slightly different wording? Having 2 subsequent paragraphs starting with _"This article details…"_ sounds a bit strange I think…

Author: I've updated it 👍 let me know if it needs any more work

of the Ember community! It's a great product and something that is well-aligned
philosophically with Ember. If you have ever had the pleasure to use Percy you
will know that it has had a lot of thought put into it and it Just Works™️ when
you start working with it.

For those of you who don't know what Percy is, on their website they describe
themselves as "Your all-in-one visual review platform". Just like Code Reviews
and Pull requests, Percy allows you to view any **visual changes** as a result
of the work that you are doing on an application. I have also heard it referred
to as a "visual regression test system", but in practice it is much more than
that. 😂

This got me thinking; if you can use Percy for visual regression tests, then why
can't you use it to do an overarching regression test for porting the technology
powering your website, sort of like a "migration regression test". After a bit
of experimentation that is what we decided to do.

## Setting up the baseline

The first time you run Percy it takes a snapshot of all pages on your site and
uses these as the "baseline" for which any subsequent change needs to be
compared against.

> > Percy doesn't always take a snapshot of your entire site; it depends on how

Reviewer: ```suggestion
Percy doesn't always take a snapshot of your entire site; it depends on how
```

> > you integrate their SDKs with your application. For the purpose of this article

Reviewer: ```suggestion
you integrate their SDKs with your application. For the purpose of this article
```

> > (and during the guides migration) we are using the "static site" command line

Reviewer: ```suggestion
(and during the guides migration) we are using the "static site" command line
```

> > interface to upload to Percy, which loops through all of your html files and

Reviewer: ```suggestion
interface to upload to Percy, which loops through all of your html files and
```

> > effectively uploads snapshots for all pages of your site.

Reviewer: ```suggestion
effectively uploads snapshots for all pages of your site.
```

Once you have this baseline loaded for your "master" branch, you can setup Percy
in your Continuous Integration (CI) service so that every Pull Request (PR) you
make will be **compared against the baseline** and you will see a visual diff of
what the site used to look like against the current changes.

![An example of the Percy interface](/assets/images/posts/2020-03-23-porting-a-site-to-ember-with-percy/percy-example.png#@920-1920)

### Getting a copy of the existing site

The first thing that we need in this process is a local copy of the existing
site we are planning to port. If you look at the [documentation for using the
Percy command line client](https://docs.percy.io/docs/command-line-client) it
tells you that you should build your static site locally and then point the CLI
client to the output folder.

As I briefly mentioned earlier, part of the issue with the old infrastructure

> for the Ember Guides is that it was so hard to contribute to that it was almost

Reviewer: ```suggestion
for the Ember Guides was that it was so hard to contribute to that it was almost
```

since it has been replaced now?

impossible to get working locally. If we can avoid having to run the
Ruby middleman setup locally, that would be a better approach, so another
way to get a static copy of the website would be to pretend that we're Google
for a few minutes and just crawl the site to download it locally.

If you are on a Linux machine you probably have `wget` installed, it is a
particularly useful tool to download single files quickly and easily. If you are
on macOS then you will probably need to use [Homebrew](https://brew.sh/) to
install it:

```bash
brew install wget
```

Before working on this project I didn't actually realise that, not only
can you use `wget` to download single files but you can also use it to crawl and
download entire websites! Here is the command that I'm using to download the
entire [Ember Website](https://emberjs.com):

```bash
wget --recursive \
     --page-requisites \
     --html-extension \
     --convert-links \
     --domains emberjs.com \
     emberjs.com
```

> > I'm running these commands on macOS with `wget` version _1.19.5_ so you might

Reviewer: ```suggestion
I'm running these commands on macOS with `wget` version _1.19.5_ so you might
```

We don't really have a styling for blockquotes currently. Also I think this isn't actually a quote anyway? :)

> > need to confirm these arguments for your platform and `wget` version

Reviewer: ```suggestion
need to confirm these arguments for your platform and `wget` version.
```

Author: I've applied all of these updates to remove the blockquotes, but pro tip: you can do multi line suggestions now 🎉 

> To add a comment on multiple lines, click and drag to select the range of lines, then click the blue comment icon.

just needed to delete them.

```
./mascots?filter=zoey.html
./mascots?filter=other.html
./mascots?filter=tomster.html
./mascots?filter=corp.html
./mascots?filter=meetup.html
./mascots?filter=conference.html
```

If you have access to the output directory of the application that you're trying
to migrate then you won't need to go through this step of cleaning up so I would
recommend that if you can.

### Uploading to Percy

It is worth referring to the instructions
https://docs.percy.io/docs/command-line-client but for us we are running the
following command:

```bash
PERCY_TOKEN=not_our_real_token PERCY_BRANCH=master percy snapshot emberjs.com
```

In this example emberjs.com is the name of the folder that was created with the
`wget` command. We need to provide the `PERCY_BRANCH` in our case because Percy
is supposed to be run in conjunction with your Version Control System (VCS) and
since we just have a folder instead of a git repository it won't autodetect our
branch. We want it to be our `master` branch because this is going to serve as
the baseline around which all of the rest of the application will be built.

> ## Setting up visual regression testing for the Ember Application

Reviewer: Is this a good heading when what you're doing is not really visual regression testing but more visual approximation or so?

Author: Updated 👍

github: mansona
twitter: real_ate
topic: ember
bio: 'Senior Software Engineer, Ember Learning Core Team member'

> description: 'Chris Manson explains how to use Percy to port a site from one technology to another, while ensuring there are no visual changes.'

Reviewer: I'd be more specific about that you're explaining how to use Percy to approximate a visual reference instead of ensuring you're not deviating from a moving reference. I think that's really the main point of the post and what makes it different from other guides to using Percy?

Author: yes that's correct, that's for sure the main point but I am finding it hard to come up with a good one liner 🤔 

How about: 
```suggestion
description: 'Chris Manson details a novel way to use Percy to create a visual baseline that you can use to confidently port a site from one technology to another.'
```

Reviewer: Or

```suggestion
description: 'Chris Manson details a novel way to use Percy for tracking approximation of a visual baseline when  porting a site from one technology to another.'
```

?

Author: got there with a bit of teamwork 🎉 😂
